internal/fs/buffer/inmemory_write_buffer_test.go

```diff
@@ -41,24 +56,370 @@ func (t *MemoryBufferTest) SetUp(ti *TestInfo) {
+}
+
+func (t *MemoryBufferTest) TestWriteRollOverFromCurrentToFlushedBuffer() {
+	sizeInMB := t.TestCreateEmptyInMemoryBufferWithValidBufferSize()
+	data1 := t.generateRandomData(MiB)
+	data2 := t.generateRandomData(2 * KiB)
+
+	// Write to buffer
+	err := t.mb.WriteAt(data1, 0)
+	AssertEq(nil, err)
+	err = t.mb.WriteAt(data2, MiB)
+	AssertEq(nil, err)
+	err = t.mb.WriteAt(data2, 2*MiB-KiB)
+	AssertEq(nil, err)
+
+	AssertEq(t.mb.fileSize, 2*MiB+KiB)
+	AssertTrue(bytes.Equal(data2[0:KiB], t.mb.flushed.buffer[MiB-KiB:MiB]))
+	AssertTrue(bytes.Equal(data2[KiB:], t.mb.current.buffer[0:KiB]))
+	t.validateInMemoryBuffer(t.mb.current, sizeInMB*MiB, 0, 2*MiB, 3*MiB)
+	t.validateInMemoryBuffer(t.mb.flushed, sizeInMB*MiB, 0, MiB, 2*MiB)
+}
+
+func (t *MemoryBufferTest) TestRandomWriteOnAnAlreadyUploadedBufferBlockShouldFail() {
+	sizeInMB := t.TestCreateEmptyInMemoryBufferWithValidBufferSize()
+	data1 := t.generateRandomData(MiB)
+	data2 := t.generateRandomData(2 * KiB)
+
+	// Write to buffer
+	err := t.mb.WriteAt(data1, 0)
+	AssertEq(nil, err)
+	err = t.mb.WriteAt(data2, MiB)
+	AssertEq(nil, err)
+	// Write to an already uploaded offset
+	err = t.mb.WriteAt([]byte(smallContent3), 0)
+
+	AssertNe(nil, err)
+	AssertEq(NonSequentialWriteError, err.Error())
+	AssertEq(t.mb.fileSize, MiB+2*KiB)
+	t.validateInMemoryBuffer(t.mb.current, sizeInMB*MiB, 0, MiB, 2*MiB)
+	t.validateInMemoryBuffer(t.mb.flushed, sizeInMB*MiB, 0, 0, MiB)
+}
+
+func (t *MemoryBufferTest) TestRandomWriteBeyondCurrentBufferAfterSequentialWritesShouldFail() {
+	sizeInMB := t.TestCreateEmptyInMemoryBufferWithValidBufferSize()
+	data1 := t.generateRandomData(MiB)
+	data2 := t.generateRandomData(2 * KiB)
+
+	// Sequential writes
+	err := t.mb.WriteAt(data1, 0)
+	AssertEq(nil, err)
+	err = t.mb.WriteAt(data2, MiB)
+	AssertEq(nil, err)
+	// Random write beyond current buffer block
+	err = t.mb.WriteAt(data2, 2*MiB+1)
+
+	AssertNe(nil, err)
+	AssertEq(NonSequentialWriteError, err.Error())
+	AssertEq(t.mb.fileSize, MiB+2*KiB)
+	t.validateInMemoryBuffer(t.mb.current, sizeInMB*MiB, 0, MiB, 2*MiB)
+	t.validateInMemoryBuffer(t.mb.flushed, sizeInMB*MiB, 0, 0, MiB)
+}
+
+func (t *MemoryBufferTest) TestSmallContentWriteWith2ChunksSkippedShouldFail() {
+	sizeInMB := t.TestCreateEmptyInMemoryBufferWithValidBufferSize()
+	data1 := t.generateRandomData(MiB)
+	data2 := t.generateRandomData(2 * KiB)
+
+	// Write to buffer sequentially
+	err := t.mb.WriteAt(data1, 0)
+	AssertEq(nil, err)
+	err = t.mb.WriteAt(data2, MiB)
+	AssertEq(nil, err)
+	// Write to an offset with 2 chunks skipped
+	err = t.mb.WriteAt([]byte(smallContent1), 5*MiB)
+
+	AssertNe(nil, err)
+	AssertEq(NonSequentialWriteError, err.Error())
+	AssertEq(t.mb.fileSize, MiB+2*KiB)
+	t.validateInMemoryBuffer(t.mb.current, sizeInMB*MiB, 0, MiB, 2*MiB)
+	t.validateInMemoryBuffer(t.mb.flushed, sizeInMB*MiB, 0, 0, MiB)
+}
+
+func (t *MemoryBufferTest) TestBigContentWriteWith2ChunksSkippedShouldFail() {
```

Comment:
isn't this case same as TestRandomWriteBeyondCurrentBufferAfterSequentialWritesShouldFail ?

Reply:
Both are hitting the same condition but are slightly different in the sense that in TestRandomWriteBeyondCurrentBufferAfterSequentialWritesShouldFail, we write in the following manner:
0-1 MiB
1 MiB - 1MiB+2Kib
// no chunk of buffer size skipped
2MiB - 3MiB (fails)

while in TestBigContentWriteWith2ChunksSkippedShouldFail, we write in the following manner:
0-1MiB
1MiB - 1MiB+2KiB
//skips two chunks of buffer size
5MiB - 7MiB - (fails)

CCing @gargnitingoogle since he suggested to add this additional test case.